qiskit/transpiler/passes/routing/sabre_swap.py

```diff
@@ -319,159 +313,114 @@ def process_dag(block_dag, wire_map):
+            qubits = (
+                physical_qubits[layout.logical_to_physical(a)],
+                physical_qubits[layout.logical_to_physical(b)],
+            )
+            layout.swap_logical(a, b)
+            dest_dag.apply_operation_back(SwapGate(), qubits, ())
 
-                # Apply the control flow gate to the dag.
-                mapped_node = node.op.replace_blocks(mapped_blocks)
-                mapped_node_qargs = mapped_blocks[0].qubits
-                out_dag.apply_operation_back(mapped_node, mapped_node_qargs, node.cargs)
-                continue
+    def recurse(dest_dag, source_dag, result, root_logical_map, layout):
+        """The main recursive worker.  Mutates ``dest_dag`` and ``layout`` and returns them.
 
-            # If we get here, the node isn't a control-flow gate.
+        ``root_virtual_map`` is a mapping of the (virtual) qubit in ``source_dag`` to the index of
+        the virtual qubit in the root source DAG that it is bound to."""
+        swap_map, node_order, node_block_results = result
+        for node_id in node_order:
+            node = source_dag._multi_graph[node_id]
             if node_id in swap_map:
-                process_swaps(
-                    swap_map[node_id],
-                    out_dag,
-                    current_layout,
-                    canonical_register,
-                    False,
-                    bit_to_qreg_idx,
-                    component_map,
+                apply_swaps(dest_dag, swap_map[node_id], layout)
+            if not isinstance(node.op, ControlFlowOp):
+                dest_dag.apply_operation_back(
+                    node.op,
+                    [
+                        physical_qubits[layout.logical_to_physical(root_logical_map[q])]
+                        for q in node.qargs
+                    ],
+                    node.cargs,
                 )
-            apply_gate(
-                out_dag,
-                node,
-                current_layout,
-                canonical_register,
-                False,
-                qubit_indices_inner,
-            )
-
-    apply_inner(mapped_dag, initial_layout, qubit_indices, sabre_result, root_dag._multi_graph)
-
-
-def process_swaps(
-    swap_list,
-    mapped_dag,
-    current_layout,
-    canonical_register,
-    fake_run,
-    qubit_indices,
-    component_map=None,
-):
-    """
-    Applies each swap in ``swap_list`` sequentially and updates
-    ``current_layout`` accordingly.
-    """
-    for swap in swap_list:
-        if component_map:
-            swap_qargs = [
-                canonical_register[component_map[swap[0]]],
-                canonical_register[component_map[swap[1]]],
-            ]
-        else:
-            swap_qargs = [canonical_register[swap[0]], canonical_register[swap[1]]]
-        apply_gate(
-            mapped_dag,
-            DAGOpNode(op=SwapGate(), qargs=swap_qargs),
```

Comment:
I'm glad to see this oddity disappear. There wasn't ever a real reason we should have been manually creating dag nodes.